netket/machine/jax.py

```diff
@@ -182,6 +187,24 @@ def parameters(self, p):
 
         assert npar == self._npar
 
+    def jacobian_flatten(self,jac):
+        r"""Flattens the jacobian to be shape [-1,n_params] 
+
+        Args:
+             data: a (possibly non-flat) structure containing jax arrays.
+
+        Returns:
+             jax.numpy.ndarray: a 2-dimensional array corresponding to the
+             Jacobian of the examples w.r.t. the parameters
+        """
+
+        return jax.numpy.concatenate(tuple(fd.reshape(len(fd),-1) for fd in tree_flatten(jac)[0]),-1)
+
+
+
+    def jax_flatten(self,data):
+        return jax.numpy.concatenate(tuple(fd.reshape(-1) for fd in tree_flatten(data)[0]))
+
     def numpy_flatten(self, data):
         r"""Returns a flattened numpy array representing the given data.
             This is typically used to serialize parameters and gradients.
```

Comment:
Looking at all the flatten/unflatten functions here: Is there a reason they a part of the machine class? They do not depend on any state of the machine (as they don't even access self) as far as I can see, and thus would make a lot more sense as free-standing functions (which is also more easily reusable, no need to create machine just to use them). Or am I missing something?

Reply:
those are methods that each machine is supposed to use to convert their internal types to flattened numpy array, so in principle each machine can have a different implementation... it's not optimal, but I was using that only for testing basically

Reply:
No reason at all, just my general coding inexperience :). I agree that these functions can be more general

Reply:
Yeah, for testing this is fine of course. I think we should do something like 1) extract the `jax_flatten` and others as utility functions to some other place, as they are probably generally useful, not just in a context where we have a machine accessible, 2) leave a pair of abstract `flatten`/`unflatten` methods on the machine, which call the appropriate utility functions for the machine type.